www/pfSense-pkg-vHosts/files/usr/local/www/vhosts_certs.php

```diff
@@ -0,0 +1,465 @@
+	/* Retrieve certs from PKCS#12 file if provided. */
+	/*-----------------------------------------------*/
+	if ($_FILES['p12file']['tmp_name']) {
+
+		$p12pass = $_POST['p12pass'] ?: '';
+		$p12path = $_FILES['p12file']['tmp_name'];
+
+		if (is_uploaded_file($p12path)) {
+			if ($p12file = file_get_contents($p12path)) {
+				if (openssl_pkcs12_read($p12file, $certs, $p12pass)) {
+					$crt = $certs[cert];
+					$key = $certs[pkey];
+				} else {
+					$input_errors[] = gettext('Import Password is incorrect or Key-Pair file is not a valid PKCS#12 file.');
+				}
+			} else {
+				$input_errors[] = gettext('Key-Pair File could not be read.');
+			}
+		} else {
+			$input_errors[] = gettext('Key-Pair File upload failed. Please try again.');
+		}
+
+	/*----------------------------------------------------------*/
+	/* Retrieve certs from user input and verify proper format. */
+	/*----------------------------------------------------------*/
+	} else {
+
+		$crt = $_POST['certcrt'];
+		$key = $_POST['certkey'];
+
+		if (empty($crt) || empty($key)) {
+			$input_errors[] = gettext('Certificate and Certificate Key are required.');
+
+		} elseif ($crt && (!strstr($crt, "BEGIN CERTIFICATE") || !strstr($crt, "END CERTIFICATE"))) {
+			$input_errors[] = gettext("This certificate does not appear to be valid.");
+		}
+	}
+
+	/*---------------------------------------*/
+	/* Verify certificate and key are valid. */
+	/*---------------------------------------*/
+	if (!$input_errors) {
+
+		/*-------------------------------------------*/
+		/* Verify cert and key belong to each other. */
+		/*-------------------------------------------*/
+		if (cert_get_modulus($crt, false) != prv_get_modulus($key, false)) {
+			$input_errors[] = gettext("The private key does not match the certificate.");
+		}
+		/*------------------------------------------*/
+		/* Verify cert is used to validate servers. */
+		/*------------------------------------------*/
+		$certinfo = openssl_x509_parse($crt);
+		$newCN    = $certinfo['subject']['CN'];
+		$purpose  = array_search('sslserver', array_column($certinfo['purposes'], 2));
+
+		if (!isset($purpose)) {
+			$input_errors[] = gettext("This certificate is not for verifying a server.");
+		}
+		/*---------------------------------*/
+		/* For an edit, verify common name */
+		/* of new cert matches the old.    */
+		/*---------------------------------*/
+		if ($good_id && $newCN != $oldcert['cn']) {
+			$input_errors[] = gettext("Common Name '$newCN' must match existing certificate Common Name.");
+		}
+
+		/*--------------------------------------*/
+		/* Verify another certificate with the  */
+		/* same common name is not in the list. */
+		/*--------------------------------------*/
+		if (array_filter($a_vcerts, function($cert) use ($newCN, $oldcert) { return ($cert['cn'] == $newCN && $cert['refid'] != $oldcert['refid']); })) {
+			$input_errors[] = gettext("A certificate with the common name '$newCN' already exists.");
```

Comment:
Safer to not print user input back, may as well just leave the variable contents out of this error message.